// @ui5/webcomponents-fiori/dist/SideNavigation.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var SideNavigation_1;
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import ResponsivePopover from "@ui5/webcomponents/dist/ResponsivePopover.js";
import event from "@ui5/webcomponents-base/dist/decorators/event.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import slot from "@ui5/webcomponents-base/dist/decorators/slot.js";
import { getI18nBundle } from "@ui5/webcomponents-base/dist/i18nBundle.js";
import ItemNavigation from "@ui5/webcomponents-base/dist/delegate/ItemNavigation.js";
import { isPhone, isTablet, isCombi, } from "@ui5/webcomponents-base/dist/Device.js";
import NavigationMode from "@ui5/webcomponents-base/dist/types/NavigationMode.js";
import Icon from "@ui5/webcomponents/dist/Icon.js";
import "@ui5/webcomponents-icons/dist/circle-task-2.js";
import "@ui5/webcomponents-icons/dist/navigation-right-arrow.js";
import "@ui5/webcomponents-icons/dist/navigation-down-arrow.js";
import SideNavigationItem from "./SideNavigationItem.js";
import SideNavigationSubItem from "./SideNavigationSubItem.js";
import SideNavigationTemplate from "./generated/templates/SideNavigationTemplate.lit.js";
import SideNavigationPopoverTemplate from "./generated/templates/SideNavigationPopoverTemplate.lit.js";
import { SIDE_NAVIGATION_POPOVER_HIDDEN_TEXT, SIDE_NAVIGATION_COLLAPSED_LIST_ARIA_ROLE_DESC, SIDE_NAVIGATION_LIST_ARIA_ROLE_DESC, } from "./generated/i18n/i18n-defaults.js";
// Styles
import SideNavigationCss from "./generated/themes/SideNavigation.css.js";
import SideNavigationPopoverCss from "./generated/themes/SideNavigationPopover.css.js";
const PAGE_UP_DOWN_SIZE = 10;
/**
 * @class
 *
 * <h3 class="comment-api-title">Overview</h3>
 *
 * The <code>SideNavigation</code> is used as a standard menu in applications.
 * It consists of three containers: header (top-aligned), main navigation section (top-aligned) and the secondary section (bottom-aligned).
 * <ul>
 * <li>The header is meant for displaying user related information - profile data, avatar, etc.</li>
 * <li>The main navigation section is related to the user’s current work context</li>
 * <li>The secondary section is mostly used to link additional information that may be of interest (legal information, developer communities, external help, contact information and so on). </li>
 * </ul>
 *
 * <h3>Usage</h3>
 *
 * Use the available <code>ui5-side-navigation-item</code> and <code>ui5-side-navigation-sub-item</code> components to build your menu.
 * The items can consist of text only or an icon with text. The use or non-use of icons must be consistent for all items on one level.
 * You must not combine entries with and without icons on the same level. We strongly recommend that you do not use icons on the second level.
 *
 * <h3>Keyboard Handling</h3>
 *
 * <h4>Fast Navigation</h4>
 * This component provides a build in fast navigation group which can be used via <code>F6 / Shift + F6</code> or <code> Ctrl + Alt(Option) + Down /  Ctrl + Alt(Option) + Up</code>.
 * In order to use this functionality, you need to import the following module:
 * <code>import "@ui5/webcomponents-base/dist/features/F6Navigation.js"</code>
 * <br><br>
 *
 * <h3>ES6 Module Import</h3>
 *
 * <code>import "@ui5/webcomponents-fiori/dist/SideNavigation.js";</code>
 * <br>
 * <code>import "@ui5/webcomponents-fiori/dist/SideNavigationItem.js";</code> (for <code>ui5-side-navigation-item</code>)
 * <br>
 * <code>import "@ui5/webcomponents-fiori/dist/SideNavigationSubItem.js";</code> (for <code>ui5-side-navigation-sub-item</code>)
 *
 * @constructor
 * @author SAP SE
 * @alias sap.ui.webc.fiori.SideNavigation
 * @extends sap.ui.webc.base.UI5Element
 * @tagname ui5-side-navigation
 * @since 1.0.0-rc.8
 * @appenddocs sap.ui.webc.fiori.SideNavigationItem sap.ui.webc.fiori.SideNavigationSubItem
 * @public
 */
let SideNavigation = SideNavigation_1 = class SideNavigation extends UI5Element {
    constructor() {
        super();
        this._flexibleItemNavigation = new ItemNavigation(this, {
            skipItemsSize: PAGE_UP_DOWN_SIZE,
            navigationMode: NavigationMode.Vertical,
            getItemsCallback: () => this.getEnabledFlexibleItems(),
        });
        this._fixedItemNavigation = new ItemNavigation(this, {
            skipItemsSize: PAGE_UP_DOWN_SIZE,
            navigationMode: NavigationMode.Vertical,
            getItemsCallback: () => this.getEnabledFixedItems(),
        });
    }
    async _onAfterOpen() {
        // as the tree/list inside the popover is never destroyed,
        // item navigation index should be managed, because items are
        // dynamically recreated and tabIndexes are not updated
        const tree = await this.getPickerTree();
        const selectedItem = tree._findSelectedItem(tree.items);
        if (selectedItem) {
            selectedItem.focus();
        }
        else {
            tree.items[0]?.focus();
        }
    }
    get accSideNavigationPopoverHiddenText() {
        return SideNavigation_1.i18nBundle.getText(SIDE_NAVIGATION_POPOVER_HIDDEN_TEXT);
    }
    get ariaRoleDescNavigationList() {
        let key = SIDE_NAVIGATION_LIST_ARIA_ROLE_DESC;
        if (this.collapsed) {
            key = SIDE_NAVIGATION_COLLAPSED_LIST_ARIA_ROLE_DESC;
        }
        return SideNavigation_1.i18nBundle.getText(key);
    }
    handlePopupItemClick(e) {
        const associatedItem = e.target.associatedItem;
        associatedItem.fireEvent("click");
        if (associatedItem.selected) {
            this.closePicker();
            return;
        }
        this._selectItem(associatedItem);
        this.closePicker();
    }
    async getPicker() {
        return (await this.getStaticAreaItemDomRef()).querySelector("[ui5-responsive-popover]");
    }
    async openPicker(opener) {
        const responsivePopover = await this.getPicker();
        responsivePopover.showAt(opener);
    }
    async closePicker() {
        const responsivePopover = await this.getPicker();
        responsivePopover.close();
    }
    async getPickerTree() {
        const picker = await this.getPicker();
        return picker.querySelector("[ui5-side-navigation]");
    }
    get hasHeader() {
        return !!this.header.length;
    }
    get showHeader() {
        return this.hasHeader && !this.collapsed;
    }
    get hasFixedItems() {
        return !!this.fixedItems.length;
    }
    get _rootRole() {
        return this._inPopover ? "none" : undefined;
    }
    get classes() {
        return {
            root: {
                "ui5-sn-phone": isPhone(),
                "ui5-sn-tablet": isTablet(),
                "ui5-sn-combi": isCombi(),
                "ui5-sn-collapsed": this.collapsed,
                "ui5-sn-in-popover": this._inPopover,
            },
        };
    }
    getEnabledFixedItems() {
        return this.getEnabledItems(this.fixedItems);
    }
    getEnabledFlexibleItems() {
        return this.getEnabledItems(this.items);
    }
    getEnabledItems(items) {
        let result = new Array();
        items.forEach(item => {
            if (!item.disabled) {
                result.push(item);
            }
            if (!this.collapsed && item.expanded) {
                result = result.concat(item.items.filter(el => !el.disabled));
            }
        });
        return result;
    }
    focusItem(item) {
        if (item.isFixedItem) {
            this._fixedItemNavigation.setCurrentItem(item);
        }
        else {
            this._flexibleItemNavigation.setCurrentItem(item);
        }
    }
    onAfterRendering() {
        const activeElement = this.shadowRoot.activeElement;
        const flexibleDom = this.shadowRoot.querySelector(".ui5-sn-flexible");
        if (!flexibleDom.contains(activeElement)) {
            const selectedItem = this._findSelectedItem(this.items);
            if (selectedItem) {
                this._flexibleItemNavigation.setCurrentItem(selectedItem);
            }
            else {
                const focusedItem = this._findFocusedItem(this.items);
                if (!focusedItem) {
                    this._flexibleItemNavigation.setCurrentItem(this.items[0]);
                }
            }
        }
        const fixedDom = this.shadowRoot.querySelector(".ui5-sn-fixed");
        if (!fixedDom?.contains(activeElement)) {
            const selectedItem = this._findSelectedItem(this.fixedItems);
            if (selectedItem) {
                this._fixedItemNavigation.setCurrentItem(selectedItem);
            }
            else {
                const focusedItem = this._findFocusedItem(this.fixedItems);
                if (!focusedItem) {
                    this._fixedItemNavigation.setCurrentItem(this.fixedItems[0]);
                }
            }
        }
    }
    _findFocusedItem(items) {
        let focusedItem;
        if (this.collapsed) {
            focusedItem = items.find(item => item._tabIndex === "0");
        }
        else {
            focusedItem = this._getWithNestedItems(items, true).find(item => item._tabIndex === "0");
        }
        return focusedItem;
    }
    _getWithNestedItems(items, expandedOnly = false) {
        let result = new Array();
        items.forEach(item => {
            result.push(item);
            if (!expandedOnly || item.expanded) {
                result = result.concat(item.items);
            }
        });
        return result;
    }
    _findSelectedItem(items) {
        let selectedItem;
        if (this.collapsed) {
            selectedItem = items.find(item => item._selected);
        }
        else {
            selectedItem = this._getWithNestedItems(items).find(current => current.selected);
        }
        return selectedItem;
    }
    _handleItemClick(e, item) {
        if (item.selected && !this.collapsed) {
            item.fireEvent("click");
            return;
        }
        if (this.collapsed && item instanceof SideNavigationItem && item.items.length) {
            e.preventDefault();
            this._popoverContents = {
                item,
                subItems: item.items,
            };
            this.openPicker(item.getFocusDomRef());
        }
        else {
            item.fireEvent("click");
            if (!item.selected) {
                this._selectItem(item);
            }
        }
    }
    _selectItem(item) {
        if (item.disabled) {
            return;
        }
        if (!this.fireEvent("selection-change", { item }, true)) {
            return;
        }
        let items = this._getWithNestedItems(this.items);
        items = items.concat(this._getWithNestedItems(this.fixedItems));
        items.forEach(current => {
            current.selected = false;
        });
        item.selected = true;
    }
    static async onDefine() {
        [SideNavigation_1.i18nBundle] = await Promise.all([
            getI18nBundle("@ui5/webcomponents-fiori"),
            super.onDefine(),
        ]);
    }
};
__decorate([
    property({ type: Boolean })
], SideNavigation.prototype, "collapsed", void 0);
__decorate([
    slot({ type: HTMLElement, invalidateOnChildChange: true, "default": true })
], SideNavigation.prototype, "items", void 0);
__decorate([
    slot()
], SideNavigation.prototype, "header", void 0);
__decorate([
    slot({ type: HTMLElement, invalidateOnChildChange: true })
], SideNavigation.prototype, "fixedItems", void 0);
__decorate([
    property({ type: Object })
], SideNavigation.prototype, "_popoverContents", void 0);
__decorate([
    property({ type: Boolean })
], SideNavigation.prototype, "_inPopover", void 0);
SideNavigation = SideNavigation_1 = __decorate([
    customElement({
        tag: "ui5-side-navigation",
        fastNavigation: true,
        renderer: litRender,
        template: SideNavigationTemplate,
        staticAreaTemplate: SideNavigationPopoverTemplate,
        styles: SideNavigationCss,
        staticAreaStyles: SideNavigationPopoverCss,
        dependencies: [
            ResponsivePopover,
            SideNavigationItem,
            SideNavigationSubItem,
            Icon,
        ],
    })
    /**
     * Fired when the selection has changed via user interaction
     *
     * @event sap.ui.webc.fiori.SideNavigation#selection-change
     * @param {sap.ui.webc.fiori.SideNavigationItemBase} item the clicked item.
     * @allowPreventDefault
     * @public
     */
    ,
    event("selection-change", {
        detail: {
            item: { type: HTMLElement },
        },
    })
], SideNavigation);
SideNavigation.define();
export default SideNavigation;
