// @ui5/webcomponents-react/wrappers/webComponents/Carousel/index.js
'use client';

import { CarouselArrowsPlacement, BackgroundDesign, BorderDesign, CarouselPageIndicatorStyle } from '../../enums/index.js';
import { withWebComponent } from '../../internal/withWebComponent.js';
/**
 * The Carousel allows the user to browse through a set of items. The component is mostly used for showing a gallery of images, but can hold any other HTML element.
 * There are several ways to perform navigation:
 *
 * *   on desktop - the user can navigate using the navigation arrows or with keyboard shorcuts.
 * *   on mobile - the user can use swipe gestures.
 *
 * __Note:__ This component is a web component developed by the UI5 Web Components’ team.
 *
 * [UI5 Web Components Storybook](https://sap.github.io/ui5-webcomponents/playground/?path=/docs/main-Carousel)
 */
const Carousel = withWebComponent('ui5-carousel', ['arrowsPlacement', 'backgroundDesign', 'itemsPerPageL', 'itemsPerPageM', 'itemsPerPageS', 'pageIndicatorBackgroundDesign', 'pageIndicatorBorderDesign', 'pageIndicatorStyle'], ['cyclic', 'hideNavigationArrows', 'hidePageIndicator'], [], ['navigate']);
Carousel.displayName = 'Carousel';
Carousel.defaultProps = {
  arrowsPlacement: CarouselArrowsPlacement.Content,
  backgroundDesign: BackgroundDesign.Translucent,
  itemsPerPageL: 1,
  itemsPerPageM: 1,
  itemsPerPageS: 1,
  pageIndicatorBackgroundDesign: BackgroundDesign.Solid,
  pageIndicatorBorderDesign: BorderDesign.Solid,
  pageIndicatorStyle: CarouselPageIndicatorStyle.Default
};
export { Carousel };
